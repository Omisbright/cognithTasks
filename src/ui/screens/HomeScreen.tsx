import React, {Component} from 'react';
import {Alert, Button, Text, TextInput, View} from 'react-native';
import {AppProps} from '../../navigation/types';
import {homeScreenStyles} from '../styles/styles';

type State = {
  asteroidId: string;
};
class HomeScreen extends Component<AppProps<'HomeScreen'>, State> {
  constructor(props: AppProps<'HomeScreen'>) {
    super(props);
    this.state = {
      asteroidId: '',
    };
  }

  handleSubmit = () => {
    const {asteroidId} = this.state;
    if (!asteroidId) {
      Alert.alert(
        'Error',
        'Please enter an asteroid ID',
        [{text: 'OK', onPress: () => console.log('OK pressed')}],
        {cancelable: false},
      );
    } else {
      this.props.navigation.navigate('AsteroidDetailsScreen', {
        asteroidId: this.state.asteroidId,
      });
    }
  };

  render() {
    return (
      <View style={homeScreenStyles.container}>
        <Text style={homeScreenStyles.text}>Enter Asteroid ID:</Text>
        <TextInput
          testID="textInput"
          style={homeScreenStyles.textInput}
          onChangeText={text => this.setState({asteroidId: text})}
          value={this.state.asteroidId}
        />
        <Button
          testID="submitButton"
          title="Submit"
          onPress={() => this.handleSubmit()}
        />
        <Button
          testID="randomAsteroidNavigateButton"
          title="Random Asteroid"
          onPress={() =>
            this.props.navigation.navigate('AsteroidDetailsScreen')
          }
        />
      </View>
    );
  }
}

export default HomeScreen;
