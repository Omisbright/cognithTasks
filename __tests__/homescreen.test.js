import React from 'react';
import renderer from 'react-test-renderer';
import HomeScreen from '../src/ui/screens/HomeScreen';
import {Alert} from 'react-native';
import {render, fireEvent} from '@testing-library/react-native';

it('renders correctly', () => {
  const tree = renderer.create(<HomeScreen />).toJSON();
  expect(tree).toMatchSnapshot();
});

it('should update state when text input changes', () => {
  const {getByTestId} = render(<HomeScreen />);
  const textInput = getByTestId('textInput');
  fireEvent.changeText(textInput, '12345');
  expect(textInput.props.value).toEqual('12345');
});

it('should navigate to AsteroidDetailsScreen when submit button is pressed', () => {
  const navigation = {navigate: jest.fn()};
  const {getByTestId} = render(<HomeScreen navigation={navigation} />);
  const textInput = getByTestId('textInput');
  const submitButton = getByTestId('submitButton');
  fireEvent.changeText(textInput, '12345');
  fireEvent.press(submitButton);
  expect(navigation.navigate).toHaveBeenCalledWith('AsteroidDetailsScreen', {
    asteroidId: '12345',
  });
});

it('should show an alert when submit button is pressed and asteroidId is empty', () => {
  const {getByTestId} = render(<HomeScreen />);
  const submitButton = getByTestId('submitButton');

  // Mock the Alert.alert function
  const alertMock = jest
    .spyOn(Alert, 'alert')
    .mockImplementation((title, message, buttons, options) => {
      expect(title).toBe('Error');
      expect(message).toBe('Please enter an asteroid ID');
      expect(buttons).toEqual([{text: 'OK', onPress: expect.any(Function)}]);
      expect(options).toEqual({cancelable: false});
    });

  fireEvent.press(submitButton);

  // Verify that the Alert.alert function was called
  expect(alertMock).toHaveBeenCalledTimes(1);

  // Restore the original Alert.alert function
  alertMock.mockRestore();
});

it('should navigate to AsteroidDetailsScreen when Random Asteroid button is pressed', () => {
  const navigation = {navigate: jest.fn()};
  const {getByTestId} = render(<HomeScreen navigation={navigation} />);

  const submitButton = getByTestId('randomAsteroidNavigateButton');
  fireEvent.press(submitButton);
  expect(navigation.navigate).toHaveBeenCalledWith('AsteroidDetailsScreen');
});
