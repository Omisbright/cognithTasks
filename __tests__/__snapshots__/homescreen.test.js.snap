// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flex": 1,
      "justifyContent": "center",
      "paddingHorizontal": 20,
    }
  }
>
  <Text
    style={
      {
        "marginBottom": 5,
      }
    }
  >
    Enter Asteroid ID:
  </Text>
  <TextInput
    onChangeText={[Function]}
    style={
      {
        "borderColor": "gray",
        "borderRadius": 5,
        "borderWidth": 1,
        "height": 40,
        "marginBottom": 20,
        "paddingLeft": 10,
        "paddingRight": 10,
        "width": "100%",
      }
    }
    testID="textInput"
    value=""
  />
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="submitButton"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Submit
      </Text>
    </View>
  </View>
  <View
    accessibilityRole="button"
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      {
        "opacity": 1,
      }
    }
    testID="randomAsteroidNavigateButton"
  >
    <View
      style={
        [
          {},
        ]
      }
    >
      <Text
        style={
          [
            {
              "color": "#007AFF",
              "fontSize": 18,
              "margin": 8,
              "textAlign": "center",
            },
          ]
        }
      >
        Random Asteroid
      </Text>
    </View>
  </View>
</View>
`;
